File: backend/app/schemas/work_queue.py
"""
Work Queue, Reminders, and Evidence Pack Schemas

Schemas for:
- Work queue items and responses
- SLA summary
- Reminder requests and responses
- Evidence pack requests and responses
"""
from datetime import datetime, date
from typing import Optional, List, Dict, Any
from uuid import UUID
from pydantic import BaseModel, Field


# ============ Work Queue Schemas ============

class WorkQueueItem(BaseModel):
    """Single work item in the queue."""
    client_id: UUID
    client_name: str
    period_id: Optional[UUID] = None
    period_status: Optional[str] = None
    work_item_type: str = Field(..., description="Type: ISSUE, VAT, BACKLOG, ALERT, PERIOD_REVIEW, STALE")
    severity: Optional[str] = Field(None, description="Severity: CRITICAL, RED, WARNING, YELLOW, INFO")
    title: str
    description: str
    suggested_next_action: str
    due_date: Optional[date] = None
    age_days: Optional[int] = None
    counts: Dict[str, int] = Field(default_factory=dict, description="Counts: red, yellow, backlog, alerts")
    readiness_score: int = Field(..., ge=0, le=100, description="Client readiness score 0-100")
    readiness_breakdown: Optional[Dict[str, Any]] = None


class WorkQueueCounts(BaseModel):
    """Summary counts for work queue."""
    red_issues: int = 0
    needs_review: int = 0
    vat_due: int = 0
    stale: int = 0


class WorkQueueResponse(BaseModel):
    """Response for GET /api/v1/accountant/work-queue."""
    items: List[WorkQueueItem]
    total_count: int
    returned_count: int
    queue_type: str
    counts: Dict[str, int]
    sort_by: str
    sort_order: str
    generated_at: datetime


# ============ SLA Summary Schemas ============

class SLAViolationByType(BaseModel):
    """Violation counts by type."""
    critical: int = 0
    warning: int = 0


class SLASummaryResponse(BaseModel):
    """Response for GET /api/v1/accountant/dashboard/sla-summary."""
    total_violations: int
    critical_count: int
    warning_count: int
    by_type: Dict[str, Dict[str, int]]
    escalation_events_today: int
    policy: Dict[str, int] = Field(..., description="Current SLA policy thresholds")
    generated_at: datetime


# ============ Reminder Schemas ============

class ReminderSendRequest(BaseModel):
    """Request for POST /api/v1/accountant/reminders/send."""
    client_ids: List[str] = Field(..., description="List of client IDs to send reminders to")
    reminder_type: str = Field(..., description="Type: DOCUMENT_MISSING, VAT_DEADLINE, REVIEW_PENDING, ACTION_REQUIRED")
    title: str = Field(..., max_length=255, description="Reminder title")
    message: str = Field(..., description="Reminder message")
    channel: str = Field("IN_APP", description="Channel: IN_APP or EMAIL")
    due_date: Optional[date] = Field(None, description="Optional due date")
    template_id: Optional[str] = Field(None, max_length=100, description="Optional email template ID")
    variables: Optional[Dict[str, Any]] = Field(None, description="Optional template variables")


class ReminderScheduleRequest(ReminderSendRequest):
    """Request for POST /api/v1/accountant/reminders/schedule."""
    scheduled_at: datetime = Field(..., description="When to send the reminders")


class ReminderResponse(BaseModel):
    """Response for a single reminder."""
    id: UUID
    administration_id: UUID
    reminder_type: str
    title: str
    message: str
    channel: str
    status: str = Field(..., description="Status: PENDING, SCHEDULED, SENT, FAILED")
    due_date: Optional[date] = None
    scheduled_at: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    send_error: Optional[str] = None

    class Config:
        from_attributes = True


class ReminderHistoryResponse(BaseModel):
    """Response for GET /api/v1/accountant/reminders/history."""
    reminders: List[ReminderResponse]
    total_count: int
    limit: int
    offset: int


# ============ Evidence Pack Schemas ============

class EvidencePackCreateRequest(BaseModel):
    """Request for POST /api/v1/accountant/clients/{client_id}/periods/{period_id}/evidence-pack."""
    pack_type: str = Field("VAT_EVIDENCE", description="Pack type: VAT_EVIDENCE or AUDIT_TRAIL")


class EvidencePackResponse(BaseModel):
    """Response for a single evidence pack."""
    id: UUID
    administration_id: UUID
    period_id: UUID
    pack_type: str
    created_at: Optional[datetime] = None
    file_size_bytes: Optional[int] = None
    checksum: str
    download_count: int = 0
    metadata: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True


class EvidencePackListResponse(BaseModel):
    """Response for GET /api/v1/accountant/evidence-packs."""
    packs: List[EvidencePackResponse]
    total_count: int
    limit: int
    offset: int

File: src/lib/api/work-queue.ts
// ============ Work Queue / Reminder / Evidence Pack APIs (types + functions) ============
// Extracted from src/lib/api.ts as part of the api.ts decomposition.
// Behavior, endpoints, and response shapes are unchanged.

import { api } from '../api'

// ============ Work Queue Summary Types ============

export interface DocumentReviewItem {
  id: string
  date: string | null
  type: string
  status: string
  vendor_customer: string | null
  amount: number | null
  link: string
}

export interface DocumentReviewSection {
  count: number
  top_items: DocumentReviewItem[]
}

export interface BankTransactionItem {
  id: string
  date: string
  description: string
  amount: number
  confidence_best_proposal: number | null
  link: string
}

export interface BankReconciliationSection {
  count: number
  top_items: BankTransactionItem[]
}

export interface VATActionsSection {
  current_period_status: string | null
  periods_needing_action_count: number
  btw_link: string
}

export interface OverdueInvoiceItem {
  id: string
  customer: string
  due_date: string
  amount: number
  link: string
}

export interface RemindersSection {
  count: number
  top_items: OverdueInvoiceItem[]
}

export interface IntegrityWarningItem {
  id: string
  severity: string
  message: string
  link: string
}

export interface IntegrityWarningsSection {
  count: number
  top_items: IntegrityWarningItem[]
}

export interface WorkQueueSummaryResponse {
  document_review: DocumentReviewSection
  bank_reconciliation: BankReconciliationSection
  vat_actions: VATActionsSection
  reminders: RemindersSection
  integrity_warnings: IntegrityWarningsSection
  generated_at: string
}

// ============ Work Queue Types ============

export interface WorkQueueCounts {
  red_issues: number
  needs_review: number
  vat_due: number
  stale: number
}

export interface ReadinessBreakdown {
  base_score: number
  deductions: Array<{
    reason: string
    count?: number
    penalty: number
    days_remaining?: number
    days_inactive?: number
  }>
  final_score: number
}

export interface WorkQueueItem {
  client_id: string
  client_name: string
  period_id: string | null
  period_status: string | null
  work_item_type: 'ISSUE' | 'VAT' | 'BACKLOG' | 'ALERT' | 'PERIOD_REVIEW' | 'STALE'
  severity: 'CRITICAL' | 'RED' | 'WARNING' | 'YELLOW' | 'INFO' | null
  title: string
  description: string
  suggested_next_action: string
  due_date: string | null
  age_days: number | null
  counts: {
    red: number
    yellow: number
    backlog: number
  }
  readiness_score: number
  readiness_breakdown: ReadinessBreakdown | null
  staleness_days?: number
}

export interface WorkQueueResponse {
  items: WorkQueueItem[]
  total_count: number
  returned_count: number
  queue_type: string
  counts: WorkQueueCounts
  sort_by: string
  sort_order: string
  generated_at: string
}

export interface SLASummaryResponse {
  total_violations: number
  critical_count: number
  warning_count: number
  by_type: Record<string, { critical: number; warning: number }>
  escalation_events_today: number
  policy: Record<string, number>
  generated_at: string
}

// ============ Reminder Types ============

export interface ReminderResponse {
  id: string
  administration_id: string
  reminder_type: string
  title: string
  message: string
  channel: 'IN_APP' | 'EMAIL'
  status: 'PENDING' | 'SCHEDULED' | 'SENT' | 'FAILED'
  due_date: string | null
  scheduled_at: string | null
  sent_at: string | null
  created_at: string | null
  send_error: string | null
}

export interface ReminderSendRequest {
  client_ids: string[]
  reminder_type: string
  title: string
  message: string
  channel?: 'IN_APP' | 'EMAIL'
  due_date?: string
  template_id?: string
  variables?: Record<string, unknown>
}

export interface ReminderScheduleRequest extends ReminderSendRequest {
  scheduled_at: string
}

export interface ReminderHistoryResponse {
  reminders: ReminderResponse[]
  total_count: number
  limit: number
  offset: number
}

// ============ Evidence Pack Types ============

export interface EvidencePackResponse {
  id: string
  administration_id: string
  period_id: string
  pack_type: 'VAT_EVIDENCE' | 'AUDIT_TRAIL'
  created_at: string | null
  file_size_bytes: number | null
  checksum: string
  download_count: number
  metadata: {
    administration_name?: string
    kvk_number?: string
    btw_number?: string
    period_name?: string
    period_status?: string
    generated_at?: string
  } | null
}

export interface EvidencePackListResponse {
  packs: EvidencePackResponse[]
  total_count: number
  limit: number
  offset: number
}

// ============ Work Queue API ============

export const workQueueApi = {
  getWorkQueue: async (
    queue?: 'red' | 'review' | 'vat_due' | 'stale' | 'all',
    limit?: number,
    sort?: string,
    order?: 'asc' | 'desc'
  ): Promise<WorkQueueResponse> => {
    const params: Record<string, unknown> = {}
    if (queue) params.queue = queue
    if (limit) params.limit = limit
    if (sort) params.sort = sort
    if (order) params.order = order
    const response = await api.get<WorkQueueResponse>('/accountant/work-queue', { params })
    return response.data
  },

  getSLASummary: async (): Promise<SLASummaryResponse> => {
    const response = await api.get<SLASummaryResponse>('/accountant/dashboard/sla-summary')
    return response.data
  },
}

// ============ Reminder API ============

export const reminderApi = {
  send: async (request: ReminderSendRequest): Promise<ReminderResponse[]> => {
    const response = await api.post<ReminderResponse[]>('/accountant/reminders/send', request)
    return response.data
  },

  schedule: async (request: ReminderScheduleRequest): Promise<ReminderResponse[]> => {
    const response = await api.post<ReminderResponse[]>('/accountant/reminders/schedule', request)
    return response.data
  },

  getHistory: async (clientId?: string, limit?: number, offset?: number): Promise<ReminderHistoryResponse> => {
    const params: Record<string, unknown> = {}
    if (clientId) params.client_id = clientId
    if (limit) params.limit = limit
    if (offset) params.offset = offset
    const response = await api.get<ReminderHistoryResponse>('/accountant/reminders/history', { params })
    return response.data
  },
}

// ============ Evidence Pack API ============

export const evidencePackApi = {
  generate: async (
    clientId: string,
    periodId: string,
    packType?: 'VAT_EVIDENCE' | 'AUDIT_TRAIL'
  ): Promise<EvidencePackResponse> => {
    const params: Record<string, unknown> = {}
    if (packType) params.pack_type = packType
    const response = await api.post<EvidencePackResponse>(
      `/accountant/clients/${clientId}/periods/${periodId}/evidence-pack`,
      {},
      { params }
    )
    return response.data
  },

  download: async (packId: string): Promise<Blob> => {
    const response = await api.get(`/accountant/evidence-packs/${packId}/download`, {
      responseType: 'blob'
    })
    return response.data
  },

  list: async (
    clientId?: string,
    periodId?: string,
    limit?: number,
    offset?: number
  ): Promise<EvidencePackListResponse> => {
    const params: Record<string, unknown> = {}
    if (clientId) params.client_id = clientId
    if (periodId) params.period_id = periodId
    if (limit) params.limit = limit
    if (offset) params.offset = offset
    const response = await api.get<EvidencePackListResponse>('/accountant/evidence-packs', { params })
    return response.data
  },
}
